celesta-sql/src/test/java/ru/curs/celesta/score/discovery/fk/FkDefaultScoreDiscoveryTest.java

```diff
@@ -0,0 +1,146 @@
+        );
+    }
+
+
+    @Test
+    @DisplayName("Score is parsed correctly when two files depend on each other")
+    void testDependencyOnEachOther() throws Exception {
+        AbstractScore s = ScoreUtil.createCelestaSqlTestScore(
+                this.getClass(),
+                "dependency_on_each_other"
+        );
+
+        Grain g = s.getGrains().get(GRAIN_NAME);
+
+        assertAll(
+                () -> assertNotNull(g),
+                () -> assertEquals(3, g.getElements(Table.class).size()),
+                () -> assertDoesNotThrow(() -> g.getElement(A_TABLE, Table.class)),
+                () -> assertDoesNotThrow(() -> g.getElement(B_TABLE, Table.class)),
+                () -> assertDoesNotThrow(() -> g.getElement(C_TABLE, Table.class))
+        );
+
+        Table a = g.getElement(A_TABLE, Table.class);
+        Table b = g.getElement(B_TABLE, Table.class);
+        Table c = g.getElement(C_TABLE, Table.class);
+
+        assertAll(
+                () -> assertEquals(1, a.getForeignKeys().size()),
+                () -> assertSame(b, a.getForeignKeys().stream().findFirst().get().getReferencedTable()),
+                () -> assertEquals(1, c.getForeignKeys().size()),
+                () -> assertSame(a, c.getForeignKeys().stream().findFirst().get().getReferencedTable())
+        );
+    }
+
+    @Test
+    @DisplayName("Score parsing fails on explicit cycle")
+    void testFailsOnExplicitCycle() {
+        assertThrows(
+                CelestaException.class,
+                () -> ScoreUtil.createCelestaSqlTestScore(
+                        this.getClass(),
+                        "explicit_cycle"
+                ));
+
+    }
+
+    @Test
+    @DisplayName("Score parsing fails on implicit cycle")
+    void testFailsOnImplicitCycle() {
+        CelestaException e = assertThrows(
+                CelestaException.class,
+                () -> ScoreUtil.createCelestaSqlTestScore(
+                        this.getClass(),
+                        "implicit_cycle"
+                ));
+
+        String expectedMessage = String.format(
+                ReferenceResolver.CYCLE_REFERENCE_ERROR_MESSAGE_TEMPLATE,
+                String.join(" -> ", "test.a", "test.b", "test.c", "test.a")
+        );
+        assertEquals(expectedMessage, e.getMessage());
+    }
+
+    @Test
+    @DisplayName("Score parsing fails on implicit deep cycle")
+    void testFailsOnDeepImplicitCycle() {
+        CelestaException e = assertThrows(
+                CelestaException.class,
+                () -> ScoreUtil.createCelestaSqlTestScore(
+                        this.getClass(),
+                        "implicit_deep_cycle"
+                ));
+
+        String expectedMessage = String.format(
+                ReferenceResolver.CYCLE_REFERENCE_ERROR_MESSAGE_TEMPLATE,
+                String.join(" -> ", "schema1.b", "schema2.c", "schema1.d", "schema1.b")
+        );
+        assertEquals(expectedMessage, e.getMessage());
+    }
+}
```

Comment:
No newline at EOF here and for some new SQL files, Checkstyle won't be happy. AFAIR we do not Checkstyle test packages, though

Reply:
Checkstyle is not interested in tests)